06-TimeManagement/timer.c

```diff
@@ -12,15 +12,109 @@
 #define MODULE_TAG			"Timer"
 #define CLASS_NAME			"Timer"
 
+static unsigned long int interval;
+module_param(interval, ulong, 0444);
+MODULE_PARM_DESC(interval, "An interval in ms for periodic output in dmesg");
+static struct timer_list print_timer;
+
+
+static void update_print_timer(unsigned long time_ms)
+{
+	int ret;
+
+	if (time_ms) {
+		ret = mod_timer(&print_timer, jiffies + msecs_to_jiffies(interval));
+		if (!ret)
+			pr_debug("%s: Timer was started\n", MODULE_TAG);
+		else
+			pr_debug("%s: Timer was updated\n", MODULE_TAG);
+	} else {
+		ret = del_timer(&print_timer);
+		if (!ret)
+			pr_debug("%s: Inactive timer was stopped\n", MODULE_TAG);
+		else
+			pr_debug("%s: Active timer was stopped\n", MODULE_TAG);
+	}
+}
+
+static void print_timer_fn(unsigned long data)
+{
+	pr_info("%s: This is a timer callback\n", MODULE_TAG);
+
+	update_print_timer(interval);
+}
+
+static ssize_t last_read_show(struct class *class, struct class_attribute *attr, char *buf)
+{
+	return 0;
+}
+
+static ssize_t last_absolute_show(struct class *class, struct class_attribute *attr, char *buf)
+{
+	return 0;
+}
+
+static ssize_t print_interval_show(struct class *class, struct class_attribute *attr, char *buf)
+{
+	return scnprintf(buf, PAGE_SIZE, "%ld\n", interval);
+}
+
+static ssize_t print_interval_store(struct class *class, struct class_attribute *attr, const char *buf, size_t count)
+{
+	unsigned long temp;
+	int ret;
+
+	ret = kstrtoul(buf, 0, &temp);
+	if (ret < 0)
+		return ret;
+	interval = temp;
+
+	update_print_timer(interval);
```

Comment:
What would be if user enter zero as interval value?

Reply:
As it was described in the topic's head a zero value allows stopping print output.